### PR TITLE
Fix Sourcegraph actions (search selection, open selection, etc)

### LINKED ISSUE #162: Disabling Cody breaks remaining Sourcegraph plug-in "Open in Sourcegraph Web" functionality
### Installation Information:

Plugin version 5.2.9816
Seen in IntelliJ

### (optional) Error message or idea.log output:

Screenshot 2023-12-04 at 4 06 42 PM

### Steps to reproduce:

1. Log in to Cody
2. Right-click a symbol and go to Sourcegraph>Open Selection in Sourcegraph Web
** Verify that a browser window opens the symbol in configured instance of Sourcegraph
3. Disable Cody by going to Settings>Sourcegraph & Cody > uncheck "Enable Cody"
5. Repeat step 2
** No browser window opens and the following error and message pop up 
Screenshot 2023-12-04 at 4 06 42 PM

5. Restart IntelliJ
6. Repeat step 2 again
** Note that no browser window opens but the error message and pop up from step 5 do not appear
7. Re-Enable Cody (following step 3 but ensure "Enable Cody" is checked)
8. Note that browser window from previous actions immediately opens

### Expected Behaviour:

Steps 5 and 6 should behave the same as 2, regardless of Cody being enabled or not. 
If there is an error, in step 8, the browser tabs should not reopen without context as some time may have passed between attempting the action and re-enabling Cody.

### LINKED ISSUE #252: bug: Open in sourcegraph (enterprise) not working with multiple errors
### Installation Information

IntelliJ IDEA 2023.2.5 (Ultimate Edition)
Build #IU-232.10227.8, built on November 9, 2023
Licensed to <redacted>
Expiration date: <redacted>
Runtime version: 17.0.9+7-b1000.46 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o.
macOS 13.6.3
GC: G1 Young Generation, G1 Old Generation
Memory: 2048M
Cores: 16
Metal Rendering is ON
Registry:
    debugger.new.tool.window.layout=true
    ide.experimental.ui=true

Non-Bundled Plugins:
    some.awesome (1.14)
    com.github.catppuccin.jetbrains (3.2.2)
    name.kropp.intellij.makefile (232.8660.88)
    org.jetbrains.plugins.ruby (232.10203.2)
    PMDPlugin (1.8.28)
    google-java-format (1.17.0.0)
    org.jetbrains.jumpToLine (0.1.14)
    thrift-syntax-fork (1.6.0)
    com.intellij.guice (232.8660.142)
    com.jetbrains.packagesearch.intellij-plugin (232.9921.28)
    com.github.camork.fileExpander (2.3)
    MavenRunHelper (4.27.1-IJ2022.2)
    Pythonid (232.10227.8)
    ru.meanmail.plugin.requirements (2022.4.1)
    com.sourcegraph.jetbrains (5.2.15221)

Kotlin: 232-1.9.0-IJ10227.8

### Describe the bug

The plugin worked fine for several months and then seemingly stopped working out of nowhere a couple of months ago. These errors happen on all of my repos

When it first stopped working, I received 2 errors:

`Error determining repository info: Failed to convert git clone URL to codebase name for cloneURL`
and
`The file is not under version control that your IDE + this plugin supports. The plugin currently only supports Git and Perforce. For the IDE part, make sure you have the Git or Perforce plugin (whichever you need) installed. If you are seeing this error for a supported VCS with the plugin installed, please reach out to [support@sourcegraph.com](mailto:support@sourcegraph.com).`

Other people on my team/org have not had the same issue, so I tried generating new tokens. When that did not work, I uninstalled and reinstalled the soursegraph plugin. After reinstall, I get the following 2 errors

`The file is not under version control that your IDE + this plugin supports ...` and `Error determining repository info: java.lang.NullPointerException: Cannot invoke "com.sourcegraph.cody.agent.CodyAgentServer.convertGitCloneURLToCodebaseName(com.sourcegraph.cody.agent.protocol.CloneURL)" because "server" is null`.

The server is set in Authentication to `https://sourcegraph.<redacted>.net/`.

```
git remote -v
origin  git@cd.<redacted>.com:<path1>/<path2>/<project>.git
```

### Expected behavior

Interacting with the sourcegraph plugin (open selection in sourcegraph web, copy file link, search) should all work without error

### Additional context (logs, images, etc)

_No response_

### LINKED ISSUE #324: bug: [RARE] `java.lang.IndexOutOfBoundsException: Index: 2, Tab count: 2` on IDE startup
```
java.lang.IndexOutOfBoundsException: Index: 2, Tab count: 2
	at java.desktop/javax.swing.JTabbedPane.checkIndex(JTabbedPane.java:1793)
	at java.desktop/javax.swing.JTabbedPane.removeTabAt(JTabbedPane.java:973)
	at java.desktop/javax.swing.JTabbedPane.remove(JTabbedPane.java:1092)
	at com.sourcegraph.cody.CodyToolWindowContent.refreshSubscriptionTab(CodyToolWindowContent.kt:160)
	at com.sourcegraph.cody.CodyToolWindowContent.access$refreshSubscriptionTab(CodyToolWindowContent.kt:44)
	at com.sourcegraph.cody.CodyToolWindowContent$refreshSubscriptionTab$1$1.invoke$lambda$0(CodyToolWindowContent.kt:141)
	at com.intellij.util.concurrency.ContextRunnable.run(ContextRunnable.java:27)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:209)
	at com.intellij.openapi.application.TransactionGuardImpl.access$100(TransactionGuardImpl.java:22)
	at com.intellij.openapi.application.TransactionGuardImpl$1.run(TransactionGuardImpl.java:191)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:843)
	at com.intellij.openapi.application.impl.ApplicationImpl$4.run(ApplicationImpl.java:463)
	at com.intellij.openapi.application.impl.RwLockHolder.runWithEnabledImplicitRead(RwLockHolder.kt:75)
	at com.intellij.openapi.application.impl.RwLockHolder.runWithImplicitRead(RwLockHolder.kt:67)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWithImplicitRead(ApplicationImpl.java:1430)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:82)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:124)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:44)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:792)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:739)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:733)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:761)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.kt:695)
	at com.intellij.ide.IdeEventQueue._dispatchEvent$lambda$12(IdeEventQueue.kt:589)
	at com.intellij.openapi.application.impl.RwLockHolder.runWithoutImplicitRead(RwLockHolder.kt:44)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:589)
	at com.intellij.ide.IdeEventQueue.access$_dispatchEvent(IdeEventQueue.kt:72)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:355)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:354)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:793)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:354)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:349)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$1(IdeEventQueue.kt:1014)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:106)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:1014)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$7(IdeEventQueue.kt:349)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:848)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.kt:391)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
```

I've seen it only once (and we do `:runIde` all the time) - I would consider this one as a minor issue.

### PR DESCRIPTION
**Pull request by mkondratek**
Fixes https://github.com/sourcegraph/jetbrains/issues/252.
Fixes https://github.com/sourcegraph/jetbrains/issues/162.
Fixes https://github.com/sourcegraph/jetbrains/issues/324.

## Test plan
- [x] test Sourcegraph actions with Cody **enabled**
- [x] test Sourcegraph actions with Cody **disabled**

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt [hunk 1]
```diff
@@ -156,7 +156,7 @@ class CodyToolWindowContent(private val project: Project) : UpdatableChat {
             SUBSCRIPTION_TAB_INDEX)
       }
       subscriptionPanel.update(data.isCurrentUserPro)
-    } else {
+    } else if (SUBSCRIPTION_TAB_INDEX < tabbedPane.tabCount) {
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/CodyToolWindowContent.kt:159]: unrel to the sourcegraph actions issue - minor fix (`IndexOutOfBoundsException` appeared once from this place)

File: src/main/java/com/sourcegraph/vcs/RepoUtil.java [hunk 2]
```diff
@@ -147,15 +151,22 @@ private static String doReplacements(
...
+
+      if (ConfigUtil.isCodyEnabled()) {
+        codebaseName =
+            CodyAgentService.getAgent(project)
+                .thenCompose(
+                    agent ->
+                        agent.getServer().convertGitCloneURLToCodebaseName(new CloneURL(cloneURL)))
+                .completeOnTimeout(/* value= */ null, /* timeout= */ 4, TimeUnit.SECONDS)
+                .get();
+      }
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/vcs/RepoUtil.java:164]: Let's try to use agent only with Cody enabled. Also let's complete on timeout.

File: src/main/java/com/sourcegraph/vcs/RepoUtil.java [hunk 3]
```diff
@@ -147,15 +151,22 @@ private static String doReplacements(
...
+
       if (codebaseName == null) {
-        throw new Exception(
-            "Failed to convert git clone URL to codebase name for cloneURL: " + cloneURL);
+        logger.warn(
+            "Failed to convert git clone URL to codebase name for cloneURL via agent: " + cloneURL);
+        codebaseName = convertGitCloneURLToCodebaseNameOrError(cloneURL);
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/vcs/RepoUtil.java:169]: This is a fallback...

File: src/main/java/com/sourcegraph/vcs/ConvertUtil.kt [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package com.sourcegraph.vcs
+
+import java.net.URL
+
+fun convertGitCloneURLToCodebaseNameOrError(cloneURL: String): String {
```
Comment (mkondratek) [inline src/main/java/com/sourcegraph/vcs/ConvertUtil.kt:5]: copied from the agent and translated (TS -> Kotlin)

**Review by pkukielka — APPROVED** [review]
LGTM

**Pull request merged**